internal/status/cache.go

```diff
@@ -61,6 +76,18 @@ func (c Cache) commitProxy(pu *ProxyUpdate) {
 		return
 	}
 
+	_, ok := c.proxyUpdates[pu.Fullname]
+	if ok {
+		// When we're committing, if we already have a Valid Condition with an error, and we're trying to
+		// set the object back to Valid, skip the commit, as we've visited too far down.
+		// If this is removed, the status reporting for when a parent delegates to a child that delegates to itself
+		// will not work. Yes, I know, problems everywhere. I'm sorry.
```

Comment:
ref an issue to clean this up.

Reply:
I'll create an issue, but I don't see how it's fixable, as it's an artifact of:
- Conditions are designed to allow multiple errors to be added, so we removed the behavior that only the first `commit()` makes it into the cache.
- If you allow multiple commits, in this case, because the child is recursed into twice during `computeRoutes`, the error is added correctly once, and then it's overwritten by a `Valid` condition.
- So, we're going to have to do something to check that we don't flip a condition back to `Valid`: `true` if there's already an error on it.

Reply:
#2968 is the related issue.

Reply:
It sounds to me like this is a problem with the API design but we can address in #2968.

> I'll create an issue, but I don't see how it's fixable, as it's an artifact of:
> 
> * Conditions are designed to allow multiple errors to be added, so we removed the behavior that only the first `commit()` makes it into the cache.

Need to consider what commit really means. If we committed a change then we want to make a change on top of that commit, the API should support that rather than hack it up at the end.

> * If you allow multiple commits, in this case, because the child is recursed into twice during `computeRoutes`, the error is added correctly once, and then it's overwritten by a `Valid` condition.

I would have thought that a child proxy was either valid or not; I'm surprised that it depends on the context. If it is dependent on context, then I expect this could also happen across proxies.

> * So, we're going to have to do something to check that we don't flip a condition back to `Valid`: `true` if there's already an error on it.

Yup, sounds like it. I think that the problem here is that `ProxyAccessor` isn't cache-aware. Each time the cache is accessed, the latest accessor should get the most recent condition state. We need to be careful about reentrancy too (either disallow that or have some merge strategy when it happens).